ArduPlane/GCS_Mavlink.cpp

```diff
@@ -794,13 +794,169 @@ MAV_RESULT GCS_MAVLINK_Plane::handle_command_int_do_reposition(const mavlink_com
     return MAV_RESULT_FAILED;
 }
 
+// these are GUIDED mode commands that are RATE or slew enabled, so you can have more powerful control than default controls.
+MAV_RESULT GCS_MAVLINK_Plane::handle_command_int_guided_slew_commands(const mavlink_command_int_t &packet)
+{
+  switch(packet.command) {
+    
+#if OFFBOARD_GUIDED == ENABLED
+    case MAV_CMD_GUIDED_CHANGE_SPEED: {
+        // command is only valid in guided mode
+        if (plane.control_mode != &plane.mode_guided) {
+            return MAV_RESULT_TEMPORARILY_REJECTED;
+        }
+
+         // only airspeed commands are supported right now...
+        if (int(packet.param1) != SPEED_TYPE_AIRSPEED) {  // since SPEED_TYPE is int in range 0-1 and packet.param1 is a *float* this works.
+            return MAV_RESULT_UNSUPPORTED;
+        }
+
+         // reject airspeeds that are outside of the tuning envelope
+        if (packet.param2 > plane.aparm.airspeed_max || packet.param2 < plane.aparm.airspeed_min) {
+            return MAV_RESULT_UNSUPPORTED;
+        }
+
+         // reject duplicate airspeeds
+        float new_target_airspeed_cm = packet.param2 * 100;
+//        if (new_target_airspeed_cm == plane.guided_state.target_airspeed_cm) {
+        if ( fabsf(new_target_airspeed_cm - plane.guided_state.target_airspeed_cm)<0.001f) { // compare if two floats are 'the same'
+            return MAV_RESULT_TEMPORARILY_REJECTED;
```

Comment:
why reject duplicates? It means if a packet is lost then the GCS will get this rejection. What does it do then? Why not accept if a tiny change? What harm does accepting it do?

Reply:
If the vehicle is commanded to the same target [one of speed,alt,direction] repeatedly, as happens often when a companion computer is streaming these packets 'repeatedly' into the autopilot, then the control subsystem has already begun heading toward that target, and is doing so with a controlled rate, so re-requesting the same target it's already going toward is both unnecessary and can change the way the algorithm performs, such as resetting deltas.  This code is basically just preventing unnecessary handing of duplicate msgs as early as possible and telling the companion computer via  MAV_RESULT_TEMPORARILY_REJECTED.   if u think this particular choice of response in this case is incorrct, i'm happy to change the response to anything that means 'we heard you, and understand u, but sorry we're not doing it" - MAV_RESULT_ACCEPTED and MAV_RESULT_UNSUPPORTED are clearly not the right response in this case so i chose MAV_RESULT_TEMPORARILY_REJECTED as it seemed appropriate.

Reply:
fixed.